readthedocsredirect.py

```diff
@@ -19,41 +19,118 @@
+    HEADERS = dict(Authorization=f"token {TOKEN}")
 
-# Change the URL
-response = requests.get(
-    f"https://readthedocs.org/api/v3/projects/{args.project}/", headers=HEADERS
-)
-assert response.ok, response.json()
-data = response.json()
-del data["language"]
-del data["programming_language"]
-data["default_branch"] = args.project
-data["repository"]["url"] = "https://github.com/dls-controls/readthedocsredirect"
-data["homepage"] = github
-response = requests.patch(
-    f"https://readthedocs.org/api/v3/projects/{args.project}/",
-    json=data,
-    headers=HEADERS,
-)
-assert response.ok, response.json()
+    def __init__(self, org: str, repo: str, project):
 
-# Setup a redirect
-response = requests.post(
-    f"https://readthedocs.org/api/v3/projects/{args.project}/redirects/",
-    json=dict(from_url="/en/latest/$rest", to_url=github, type="exact"),
-    headers=HEADERS,
-)
-assert response.ok, response.json()
+        self.project = project
+        self.github_url = f"https://{org}.github.io/{repo}/master/"
+        self.project_url = f"https://readthedocs.org/api/v3/projects/{project}"
+        self.read_the_docs_url = f"https://{project}.readthedocs.io"
+
+    def link_project_to_github(self):
+        response = requests.get(
+            self.project_url, headers=self.HEADERS
+        )
+        assert response.ok, response.json()
+
+        data = response.json()
+        data.update(
+            {
+            "default_branch": self.project,
+            "repository": {"url":  "https://github.com/dls-controls/readthedocsredirect"},
+            "homepage": self.github_url,
+            }
+        )
+        response = requests.patch(
+            self.project_url,
+            json=data,
+            headers=self.HEADERS,
+        )
+        assert response.ok, response.json()
+
+    def check_redirects(self) -> Set[str]:
+        """
+        Checks for redirects already set up through this script.
+        Errors if redirects are present which aren't expected
+        """
+
+        response = requests.get(
+            f"{self.project_url}/redirects/",
+            headers=self.HEADERS
+        )
+        assert response.ok, response.json()
+
+        already_redirected = set()
+
+        for redirect in response.json()["results"]:
+            if redirect["to_url"] != self.github_url:
+                raise RuntimeError(
+                    "Project contains an unknown redirect "
+                    f"{redirect['from_url']} to {redirect['to_url']}."
+                    "Delete it before proceeding."
+                )
+            already_redirected.add(redirect["from_url"].split("/")[2])
+
+        return already_redirected
+
+    def get_active_versions(self) -> Set[str]:
+        # Set a redirect on all active versions
+        response = requests.get(
+            f"{self.project_url}/versions", headers=self.HEADERS
+        )
+        assert response.ok, response.json()
```

Comment:
I recall you mentioned that this assert wasn't failing quite correctly? Has that been resolved?

Reply:
It was because:

``` python
        assert response.ok, response.json()
        print(response.json())
```
was failing on the second `response.json()`

for the `responses.patch` call.

This is expected because patch doesn't return a json, and the assertion passed based on the fact response.ok was True.

I've actually noticed I don't need to send across all the data received, only patch in data - I'll push this now.